library/common/engine.cc

```diff
@@ -70,123 +75,130 @@ envoy_status_t Engine::run(const std::string config, const std::string log_level
+
+    postinit_callback_handler_ = main_common->server()->lifecycleNotifier().registerCallback(
         Envoy::Server::ServerLifecycleNotifier::Stage::PostInit, [this]() -> void {
-          server_ = TS_UNCHECKED_READ(main_common_)->server();
           client_scope_ = server_->serverFactoryContext().scope().createScope("pulse.");
           // StatNameSet is lock-free, the benefit of using it is being able to create StatsName
           // on-the-fly without risking contention on system with lots of threads.
           // It also comes with ease of programming.
           stat_name_set_ = client_scope_->symbolTable().makeSet("pulse");
           auto api_listener = server_->listenerManager().apiListener()->get().http();
           ASSERT(api_listener.has_value());
-          http_dispatcher_->ready(server_->dispatcher(), server_->serverFactoryContext().scope(),
-                                  api_listener.value());
+          http_client_ = std::make_unique<Http::Client>(api_listener.value(), *dispatcher_,
+                                                        server_->serverFactoryContext().scope(),
+                                                        preferred_network_);
+          dispatcher_->drain(server_->dispatcher());
           if (callbacks_.on_engine_running != nullptr) {
             callbacks_.on_engine_running(callbacks_.context);
           }
         });
   } // mutex_
 
   // The main run loop must run without holding the mutex, so that the destructor can acquire it.
-  bool run_success = TS_UNCHECKED_READ(main_common_)->run();
+  bool run_success = main_common->run();
   // The above call is blocking; at this point the event loop has exited.
 
   // Ensure destructors run on Envoy's main thread.
   postinit_callback_handler_.reset(nullptr);
   client_scope_.reset(nullptr);
   stat_name_set_.reset();
   lambda_logger_.reset(nullptr);
-  TS_UNCHECKED_READ(main_common_).reset(nullptr);
+  main_common.reset(nullptr);
 
   callbacks_.on_exit(callbacks_.context);
 
   return run_success ? ENVOY_SUCCESS : ENVOY_FAILURE;
 }
 
-Engine::~Engine() {
-  // If we're already on the main thread, it should be safe to simply destruct.
+envoy_status_t Engine::terminate() {
+  // If main_thread_ has finished (or hasn't started), there's nothing more to do.
   if (!main_thread_.joinable()) {
-    return;
+    return ENVOY_FAILURE;
   }
 
-  // If we're not on the main thread, we need to be sure that MainCommon is finished being
-  // constructed so we can dispatch shutdown.
+  // We need to be sure that MainCommon is finished being constructed so we can dispatch shutdown.
   {
     Thread::LockGuard lock(mutex_);
 
-    if (!main_common_) {
+    if (!event_dispatcher_) {
       cv_.wait(mutex_);
     }
 
-    ASSERT(main_common_);
+    ASSERT(event_dispatcher_);
 
     // Exit the event loop and finish up in Engine::run(...)
-    event_dispatcher_->exit();
-  } // _mutex
+    if (std::this_thread::get_id() == main_thread_.get_id()) {
```

Comment:
Didn't we want to clean this to a RELEASE_ASSERT?

Reply:
I'd prefer to leave it like this for now since I want to figure out the alternate path.

Reply:
I don't think this panic will log properly. This is the main reason I am pushing for RELEASE_ASSERT. Or even better ENVOY_BUG. tioli